include/counter.h

```diff
@@ -100,26 +250,85 @@ static inline u32_t _impl_counter_read(struct device *dev)
- * @param callback Pointer to the callback function. If this is NULL,
- *                 this function unsets the alarm.
- * @param count Number of counter ticks. This is relative value, meaning
- *              an alarm will be triggered count ticks in the future.
- * @param user_data Pointer to user data.
+ *
+ * In case of absolute request, maximal value that can be set is equal to
+ * wrap value set by counter_set_wrap call or default, maximal one. in case of
+ * relative request, Maximal value can be retrieved using
+ * counter_get_max_relative_alarm.
+ *
+ * @param dev		Pointer to the device structure for the driver instance.
+ * @param alarm_cfg	Alarm configuration.
  *
  * @retval 0 If successful.
- * @retval -ENOTSUP if the counter was not started yet.
- * @retval -ENODEV if the device doesn't support interrupt (e.g. free
- *                        running counters).
- * @retval Negative errno code if failure.
+ * @retval -ENOTSUP if request is not supported (device does not support
+ *		    interrupts or requested channel).
+ * @retval -EINVAL if alarm settings are invalid.
  */
 static inline int counter_set_alarm(struct device *dev,
-				    counter_callback_t callback,
-				    u32_t count, void *user_data)
+				    const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct counter_driver_api *api = dev->driver_api;
+
+	if (alarm_cfg->channel_id >= counter_get_num_of_channels(dev)) {
+		return -ENOTSUP;
+	}
+
+	return api->set_alarm(dev, alarm_cfg);
+}
+
+/**
+ * @brief Disable an alarm.
+ *
+ * @param dev		Pointer to the device structure for the driver instance.
+ * @param alarm_cfg	Alarm configuration. It must be the same address as the
+ *			one used for @ref counter_set_alarm.
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP if request is not supported or the counter was not started
+ *		    yet.
+ */
+static inline int counter_disable_alarm(struct device *dev,
+				const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct counter_driver_api *api = dev->driver_api;
+
+	if (alarm_cfg->channel_id >= counter_get_num_of_channels(dev)) {
+		return -ENOTSUP;
+	}
+
+	return api->disable_alarm(dev, alarm_cfg);
+}
+
+/**
+ * @brief Set alarm on counter wrap.
+ *
+ * Function sets wrap value and resets the counter to 0 or wrap value
+ * depending on counter direction. On wrap, callback is periodically called and
+ * counter is reset. Wrap value can only be changed when there is no active
+ * alarm.
+ *
+ * @param dev		Pointer to the device structure for the driver instance.
+ * @param ticks		Wrap value.
+ * @param callback	Callback function. Can be NULL.
+ * @param user_data	User data passed to callback function. Not valid if
+ *			callback is NULL.
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP if request is not supported (e.g. wrap value cannot be
+ *		    changed).
+ * @retval -EBUSY if any alarm is active.
+ */
+static inline int counter_set_wrap_alarm(struct device *dev,
```

Comment:
After having another look at this I started wondering if it wouldn't be better to name this function `counter_set_wrap`, since the callback, hence the alarm, is optional.

Reply:
done